File: lib/inspector_accessibility/domain/models/ui_node.dart
import 'package:flutter/material.dart';
import 'node_selector.dart';

class UiNode {
  final String id;
  final NodeSelector selector;
  final Rect bounds;
  final String className;
  final String packageName;
  final String? viewIdResourceName;
  final bool clickable;
  final bool enabled;
  final bool scrollable;
  final bool isTextField;
  final String? text;

  UiNode({
    required this.id,
    required this.selector,
    required this.bounds,
    required this.className,
    required this.packageName,
    this.viewIdResourceName,
    required this.clickable,
    required this.enabled,
    required this.scrollable,
    required this.isTextField,
    this.text,
  });

  factory UiNode.fromJson(Map<String, dynamic> json) {
    final boundsJson = json['bounds'] as Map<String, dynamic>;
    return UiNode(
      id: json['id'] as String,
      selector: NodeSelector.fromJson(json['selector'] as Map<String, dynamic>),
      bounds: Rect.fromLTRB(
        (boundsJson['left'] as num).toDouble(),
        (boundsJson['top'] as num).toDouble(),
        (boundsJson['right'] as num).toDouble(),
        (boundsJson['bottom'] as num).toDouble(),
      ),
      className: json['className'] as String,
      packageName: json['packageName'] as String,
      viewIdResourceName: json['viewIdResourceName'] as String?,
      clickable: json['clickable'] as bool,
      enabled: json['enabled'] as bool,
      scrollable: json['scrollable'] as bool,
      isTextField: json['isTextField'] as bool,
      text: json['text'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'selector': selector.toJson(),
      'bounds': {
        'left': bounds.left,
        'top': bounds.top,
        'right': bounds.right,
        'bottom': bounds.bottom,
      },
      'className': className,
      'packageName': packageName,
      if (viewIdResourceName != null) 'viewIdResourceName': viewIdResourceName,
      'clickable': clickable,
      'enabled': enabled,
      'scrollable': scrollable,
      'isTextField': isTextField,
      if (text != null) 'text': text,
    };
  }
}

File: android/app/src/main/kotlin/com/example/neo/UiNode.kt
package com.example.neo

import android.graphics.Rect
import org.json.JSONObject

data class NodeSelector(
    val viewId: String?,
    val className: String,
    val path: String?
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            if (viewId != null) put("viewId", viewId)
            put("className", className)
            if (path != null) put("path", path)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): NodeSelector {
            return NodeSelector(
                viewId = if (json.has("viewId")) json.getString("viewId") else null,
                className = json.getString("className"),
                path = if (json.has("path")) json.getString("path") else null
            )
        }
    }

    override fun toString(): String {
        return buildString {
            if (viewId != null) append("viewId=$viewId, ")
            append("class=$className")
            if (path != null) append(", path=$path")
        }
    }
}

data class UiNode(
    val id: String,
    val selector: NodeSelector,
    val bounds: Rect,
    val className: String,
    val packageName: String,
    val viewIdResourceName: String?,
    val clickable: Boolean,
    val enabled: Boolean,
    val scrollable: Boolean,
    val isTextField: Boolean,
    val text: String? = null
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("id", id)
            put("selector", selector.toJson())
            put("bounds", JSONObject().apply {
                put("left", bounds.left)
                put("top", bounds.top)
                put("right", bounds.right)
                put("bottom", bounds.bottom)
            })
            put("className", className)
            put("packageName", packageName)
            if (viewIdResourceName != null) put("viewIdResourceName", viewIdResourceName)
            put("clickable", clickable)
            put("enabled", enabled)
            put("scrollable", scrollable)
            put("isTextField", isTextField)
            if (text != null) put("text", text)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): UiNode {
            val boundsJson = json.getJSONObject("bounds")
            val bounds = Rect(
                boundsJson.getInt("left"),
                boundsJson.getInt("top"),
                boundsJson.getInt("right"),
                boundsJson.getInt("bottom")
            )
            return UiNode(
                id = json.getString("id"),
                selector = NodeSelector.fromJson(json.getJSONObject("selector")),
                bounds = bounds,
                className = json.getString("className"),
                packageName = json.getString("packageName"),
                viewIdResourceName = if (json.has("viewIdResourceName")) json.getString("viewIdResourceName") else null,
                clickable = json.getBoolean("clickable"),
                enabled = json.getBoolean("enabled"),
                scrollable = json.getBoolean("scrollable"),
                isTextField = json.getBoolean("isTextField"),
                text = if (json.has("text")) json.getString("text") else null
            )
        }
    }
}

